MyHashMap.java

```diff
@@ -0,0 +1,196 @@
+            if (capacity >= threshold) {
+                resize();
+            }
+            putVal(hashIndex(key, nodesArr.length), value, key);
+            return value;
+        }
+        return null;
+    }
+
+    public int hashIndex(K key, int n) {
+        if (key == null) {
+            return 0;
+        }
+        return hashCode(key) & (n - 1);
+    }
+
+    public int hashCode(K key) {
+        return Objects.hashCode(key);
+    }
+
+    private void putVal(int index, V value, K key) {
+        if (isEmpty(nodesArr[index])) {
+            nodesArr[index] = new Node(hashCode(key), key, value, null);
+        } else {
+            Node currentNode = nodesArr[index];
+            while (!isEmpty(currentNode.next)) {
+                currentNode = currentNode.next;
+            }
+            currentNode.next = new Node(hashCode(key), key, value, null);
+        }
+    }
+
+    private void resize() {
+        int size = nodesArr.length;
+        Node[] tempNode = new MyHashMap.Node[size * 2];
+        System.arraycopy(nodesArr, 0, tempNode, 0, size);
+        this.nodesArr = tempNode;
+        this.size = size;
+        this.threshold = (int) (nodesArr.length * loadFactor);
+    }
+
+    @Override
+    public V remove(K key) {
+        V val = null;
+        if (key != null) {
+            int sizeArr = size;
+            while (sizeArr <= nodesArr.length) {
+                int index = hashIndex(key, sizeArr);
+                val = removeVal(index, key);
+                sizeArr *= 2;
+            }
+        }
+        return val;
+    }
+
+    public V removeVal(int index, K key) {
+        if (!isEmpty(nodesArr[index])) {
+            if (nodesArr[index].key.equals(key)) {
+                Node currNode = nodesArr[index];
+                nodesArr[index] = nodesArr[index].next;
+                return currNode.value;
+            } else {
+                Node lastNode = nodesArr[index];
+                Node currNode = nodesArr[index].next;
+                if (currNode.key.equals(key)) {
+                    lastNode.next = currNode.next;
+                    return currNode.value;
+                }
+                while (!isEmpty(currNode.next)) {
+                    lastNode = currNode;
+                    currNode = lastNode.next;
+                    if (nodesArr[index].key.equals(key)) {
+                        lastNode.next = currNode.next;
+                        return currNode.value;
+                    }
+                }
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public void clear() {
+        threshold = (int) (loadFactor * size);
+        capacity = 0;
+        nodesArr = new MyHashMap.Node[size];
+    }
+
+    @Override
+    public int size() {
+        return nodesArr.length;
+    }
+
+    @Override
+    public V get(K key) {
+        V val = null;
+        if (key != null) {
+            int sizeArr = size;
+            while (sizeArr <= nodesArr.length) {

```

Comment:
why do you need this while?  just calculate index by hashCode and check if  your bucket contains key

Reply:
> why do you need this while? just calculate index by hashCode and check if your bucket contains key

hash index depends on the size, that is, if for the current size, hashindex is already different for the same key, you need to check all the sizes, maybe there is a better way, but I did not find information about it on the Internet and did not come up with myself

Reply:
index of bucket is Math.abs(hashCode(key)) % nodesArr.length. just get node by this index and find node with right key if there multiple nodes in bucket.  All your operations should access/remove/modify nodes in this way. When you will add node and you need to make table bigger, just put all data to new array, but be aware that your nodesArr.length will be another so all nodes may have new indexes